wait for network after resuming from suspend

verybadsoldier opened the pull request:
Makes Kodi wait for network connection after resuming from suspend.

## Description
This introduces a function `WaitForNet` in Kodi's power manager which will block (with a timeout) until at least one network interface is initialized and connected. This function is invoked in `OnWake` so it is triggered on every wake up.
This behavior is disabled by default and has to be activated by a new advanced setting:
```
    <suspend>
        <waitnetonwakeup>true</waitnetonwakeup>
    </suspend>
```

This new setting would have to be documented on the Wiki I guess.

## Motivation and Context
When waking the machine from sleep/suspend and directly trying to access network resources (e.g. playing a video or reading from the MySQL DB) then Kodi will fail cause the network has not been initialized yet.
I have experienced and described this issue here:
http://forum.kodi.tv/showthread.php?tid=301976

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
I tested it on a Intel NUC i5 using LibreELEC 7.02 using a wired network. I tested it by suspending the machine and waking it up again. Then I observed the behavior with and without this setting enabled. With the setting enabled Kodi blocked for several seconds and put something like this in the logfile:

```
12:51:10 T:140441393608768  NOTICE: OnWake: Running resume jobs
12:51:10 T:140441393608768  NOTICE: WaitForNet: Waiting for first network interface to come up
12:51:18 T:140441393608768  NOTICE: WaitForNet: network interface 'eth0' is up after waiting 7400 ms
12:51:18 T:140441393608768  NOTICE: OnWake: Restarting lirc
```

After this I could use Kodi and directly access network resources without issues.

## Types of change
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

Commit:
wait for network after waking up

FernetMenta commented:
no advanced settings. those are only valuable for the very few who know about them. mostly only the developer who implemented it.

verybadsoldier commented:
@FernetMenta 
Okay, so it should be a regular setting configurable in the UI? I will try to find out how to do this.

So the general concept is accetable and might get merged if the setting is changed?

FrankHeimes commented on xbmc/powermanagement/PowerManager.cpp:
```diff
@@ -327,3 +333,25 @@ void CPowerManager::SettingOptionsShutdownStatesFiller(const CSetting *setting,
+
+  for(unsigned i=0; i < maxLoopCount; ++i)
+  {
+    CNetworkInterface* pIface = g_application.getNetwork().GetFirstConnectedInterface();
```
Consider replacing `CNetworkInterface*` with `const auto * const` if `GetName()` is `const` or `const auto` otherwise.

FrankHeimes commented on xbmc/powermanagement/PowerManager.h:
```diff
@@ -93,6 +93,8 @@ class CPowerManager : public IPowerEventsCallback
 
   void OnLowBattery();
 
+  void WaitForNet();
```
Consider declaring `WaitForNet() const`.

FrankHeimes commented on xbmc/settings/AdvancedSettings.h:
```diff
@@ -383,6 +383,8 @@ class CAdvancedSettings : public ISettingCallback, public ISettingsHandler
 
     std::string m_userAgent;
 
+    bool m_waitForNetAfterWakeUp;
```
Consider initializing member variable when declaring it: `bool m_waitForNetAfterWakeUp = true;`

FrankHeimes commented on xbmc/settings/AdvancedSettings.cpp:
```diff
@@ -1199,6 +1199,12 @@ void CAdvancedSettings::ParseSettingsFile(const std::string &file)
+  if (pElement)
+  {
+    XMLUtils::GetBoolean(pElement, "waitnetonwakeup", m_waitForNetAfterWakeUp);
+  }
```
`CAdvancedSettings::ParseSettingsFile()` is ridiculously long and should be refactored into local free functions.
That is too much work for a single feature implementation, but not adding more inline code would be a start.

Consider putting lines 1202 - 1206 into a free function
```
namespace {
	bool GetWaitNetOnWakeupFlag(TiXmlElement *rootElement)
	{
		if (!rootElement)
			throw std::invalid_argument("rootElement");

		auto firstChild = rootElement->FirstChildElement("suspend");
		if (!firstChild)
			return true; // Default behavior is to enable waitnetonwakeup

		bool waitForNetAfterWakeUp = true;
		XMLUtils::GetBoolean(firstChild, "waitnetonwakeup", waitForNetAfterWakeUp);
		return waitForNetAfterWakeUp;
	}
}
```
The line `1202: m_waitForNetAfterWakeUp = GetWaitNetOnWakeupFlag();` makes clear that the member is always well defined. Now, this is a lot more code than the original, but it is self-contained. `pRootElement` is checked for NULL, but this happens in line 444, i.e. 750+ lines above. Any (future changes) in between could invalidate it.

BTW.: IMHO, `XMLUtils` should provide function-style Getters, too, that don't require handing over a reference variable.

Yes, I'm new to this group and I do the code review like I do at work. Am I pushing too hard?

MartijnKaijser replied:
We welcome constructive reviews 👍

Jalle19 replied:
I thought you were a bot at first since the comments were so methodical and clean. +1

FernetMenta commented:
@verybadsoldier for what reason should we wait not always for network to be ready?

verybadsoldier commented:
@FernetMenta
For me it would be fine if it was just hardwired but I am afraid it could be a problem for systems without network (e.g. no NIC at all or no cable connected). Also there might be users who run Kodi exclusively from local storage and do not care for network connections. 
In both cases an unwanted delay would be introduced on wakeup. In my case it takes about 7 seconds for the network to beccome available.

verybadsoldier commented:
Reworked so that this only kicks in if there is at least one network interface. This will prevent waiting for network when there are no network Interfaces in the system. So maybe I really could just remove the advanced setting and make waiting for network after wakeup the standard behavior.

MartijnKaijser commented:
Regardless an advancedsettings is out of the question here

t4-ravenbird commented:
The original problem that this PR seeks to solve is already possible to fix by enabling the built-in [wake-on-lan](http://kodi.wiki/view/wake%20on%20lan)

That solution works the same when starting up, and when resuming from suspend. There have been MySQL issues in both cases and they are both solved when using wake-on-lan, for instance ;
http://forum.kodi.tv/showthread.php?tid=213321&pid=1879249#pid1879249
http://forum.kodi.tv/showthread.php?tid=173267&pid=1820808#pid1820808

verybadsoldier commented:
Okay, removed the advanced settings.

@t4-ravenbird 
Yes, the wake-on-lan feature can help here but to be honest I see it only as a workaround:
- the original problem is not related to WOL, so having to enable the WOL-feature feels a bit "hackish"
- the WOL-wait does not really wait until the network connection is established, according to the code it seems to just wait the configured period of time regardless of the network state
- it only kicks in when the network is actually accessed (which makes sense in the WOL context), but in my opinion Kodi's network itself should be up and working after resuming

t4-ravenbird commented:
WOL-wait waits until the correct network interface is available before proceeding (As opposed to just waiting for 1st interface)
Any thread accessing the network resource is held back until it is ready - any other threads are free to run.

verybadsoldier commented:
I fiddled around lots with this WOL-wait today and it does not really fix the problem. In some cases it did help but in most cases it did not get triggered when accessing network resources after resuming.

FernetMenta commented on xbmc/powermanagement/PowerManager.cpp:
```diff
@@ -327,3 +331,30 @@ void CPowerManager::SettingOptionsShutdownStatesFiller(const CSetting *setting,
+
+  CLog::Log(LOGNOTICE, "%s: Waiting for a network interface to come up", __FUNCTION__);
+
+  const static int numMaxTries = 50;
```
those are kind of ugly. I would make them configurable as max waiting time. a value of zero disables this feature.

verybadsoldier commented:
Agreed, I changed the loop and also added a setting which is configurable in the UI under "System -> Power saving".

![image](https://cloud.githubusercontent.com/assets/364469/21571271/f7a9bcc2-cecd-11e6-906c-c4d6786c9f05.png)

MartijnKaijser commented:
I think a max of 120 seconds is a bit much. By that time i would have throw it out the window. 30 might be more realistic although still a lot

stefansaraev commented:
there is an unwanted mode (0644 -> 755) change in this PR. please chmod everything back to 0644

verybadsoldier commented:
Reduced maximum timeout value to 30.

Thanks for pointing out the mode changes, fixed. Sorry for that.

FernetMenta commented:
looks good to me
thanks!

MartijnKaijser commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -20394,3 +20394,14 @@ msgstr ""
+#. Description of setting with label #39011 "Wait for network after waking up"
+#: system/settings/settings.xml
+msgctxt "#39012"
+msgid "Set the maximum time to wait for the network to come up after waking up."
```
"Set the maximum time to wait for the network to come up after starting or waking up. When time has past before network is up startup will continue."
If it indeed also handles a startup (which it should imo)

MartijnKaijser commented:
See my comments about the text. Once those are done we can let it build and merge it in.

chewitt commented:
@verybadsoldier Does this also handle "wait on startup" ?? if not, could it do both? There's been a long-standing issue for MySQL users where Kodi starts and the network stack isn't ready. LE has a delayed start workaround function that works, but it would be cleaner if Kodi was able to handle it internally.

verybadsoldier commented:
Ok, I have changed the wording, thanks for the feedback.

Currently it does only handle the wait when waking up. I will take a look to reuse this feature also on startup. But if possible I would like to finish this first and then do it in a separate PR if that would be ok for. I promise to come back :)

MartijnKaijser commented:
👍 might want to remove "starting or" part of the strin then and we're ready to go

verybadsoldier commented:
Oh, good point, fixed :D Thanks!

MartijnKaijser commented:
jenkins build this please

MartijnKaijser commented:
Jenkins build and merge

MartijnKaijser commented:
Jenkins build this please

verybadsoldier commented:
(Hopefully) fixed header issue for Android/iOS. Would you please trigger Jenkins again to see if it works?
Sorry I cannot test this on my end :(

MartijnKaijser commented:
No problem. That why we have jenkins build this please
